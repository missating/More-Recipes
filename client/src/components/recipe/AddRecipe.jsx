import React from 'react';
import axios from 'axios';
import Dropzone from 'react-dropzone';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';

// validations
import recipeValidator from '../../validation/recipeValidator';

// action
import addRecipe from '../../actions/addRecipe';


/**
 * @description Creates AddRecipe component
 *
 * @class AddRecipe
 *
 * @extends {React.Component}
 */
export class AddRecipe extends React.Component {
  /**
   * @description Creates an instance of AddRecipe.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      newRecipe: {
        name: '',
        ingredients: '',
        description: '',
        recipeImage: '',
        submitting: false
      },
      errors: {},
      // eslint-disable-next-line
      defaultImg: 'https://res.cloudinary.com/dxayftnxb/image/upload/v1517914951/noImage_u3sry1.png',
    };

    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
  }

  /**
   * @description Bind the value of the inputs to state
   *
   * @method onChange
   *
   * @param {object} event
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   */
  onChange(event) {
    this.setState({
      newRecipe: {
        ...this.state.newRecipe,
        [event.target.name]: event.target.value
      }
    });
  }

  /**
   * @description Submit the form
   *
   * @method onSubmit
   *
   * @param {object} event
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   */
  onSubmit(event) {
    event.preventDefault();
    this.setState({
      submitting: true
    });
    const { newRecipe } = this.state;
    const isValid = this.isValid();
    if (isValid) {
      if (newRecipe.recipeImage) {
        this.uploadToCloudinary().then((response) => {
          const secureUrl = response.data.secure_url;
          const recipeData = this.state.newRecipe;
          recipeData.recipeImage = secureUrl;

          this.props.addNewRecipe(recipeData)
            .then(() => {
              this.setState({
                newRecipe: {
                  name: '',
                  ingredients: '',
                  description: '',
                  recipeImage: '',
                  submitting: false,
                }
              });
              this.props.history.push('/users/recipes');
            });
        });
      } else {
        const recipeData = { ...newRecipe, recipeImage: this.state.defaultImg };
        this.props.addNewRecipe(recipeData)
          .then(() => {
            this.setState({
              newRecipe: {
                name: '',
                ingredients: '',
                description: '',
                recipeImage: '',
                submitting: false,
              }
            });
            this.props.history.push('/users/recipes');
          });
      }
    }
  }

  /**
   * @description Handles image drop
   *
   * @method handleDrop
   *
   * @memberof AddRecipe
   *
   * @param {array} files
   *
   * @returns {void}
   */
  handleDrop(files) {
    this.setState({
      newRecipe: {
        recipeImage: files[0]
      }
    });
  }


  /**
   * @description Handle image upload
   *
   * @method  uploadToCloudinary
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   */
  uploadToCloudinary() {
    const formData = new FormData();
    formData.append('file', this.state.newRecipe.recipeImage);
    formData.append('upload_preset', 'moreRecipes');
    formData.append('api_key', '462882972372719');

    return axios.post(
      'https://api.cloudinary.com/v1_1/dxayftnxb/image/upload',
      formData, {
        headers: { 'X-Requested-With': 'XMLHttpRequest' },
      }
    );
  }

  /**
   * @description Validates user's data before making post request
   *
   * @method isValid
   *
   * @memberof AddRecipe
   *
   * @returns {boolean} true or false
   */
  isValid() {
    const { errors, isValid } = recipeValidator(this.state.newRecipe);
    if (!isValid) {
      this.setState({ errors, submitting: false });
    } else {
      this.setState({ errors: {} });
      return isValid;
    }
  }

  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof AddRecipe
   *
   * @returns {void}
   *
   */
  render() {
    const { errors, newRecipe, defaultImg } = this.state;
    return (
      <div>
        {
          !this.props.authenticated &&
          <Redirect to="/" />
        }


        <section className="container" id="recipes">

          <h3 className="text-center mb-4">New Recipe</h3>

          <div className="row">
            <div className="col-md-12 col-sm-12">
              <form onSubmit={this.onSubmit}>
                <Dropzone
                  onDrop={this.handleDrop}
                  accept="image/*"
                  multiple={false}
                  style={{
                    border: '3px dashed ',
                    width: '300px',
                    height: 'auto',
                    margin: '0 auto',
                    padding: '5px'
                  }}
                >
                  {!newRecipe.recipeImage.preview &&
                    <div>
                      <img
                        src={defaultImg}
                        alt=""
                        className="img-fluid mx-auto d-block"
                      />
                      <h5 className="text-center">
                        Click here to upload
                      </h5>
                    </div>
                  }
                  {
                    newRecipe.recipeImage.preview &&
                    <img
                      src={newRecipe.recipeImage.preview}
                      alt=""
                      className="img-fluid mx-auto d-block"
                    />
                  }
                </Dropzone>

                <div className="form-group form-width">
                  <input
                    type="text"
                    id="recipeName"
                    name="name"
                    className="form-control"
                    placeholder="The name of the dish..."
                    value={newRecipe.name}
                    onChange={this.onChange}
                  />
                  {
                    errors.name &&
                    <span
                      className="help-block text-danger"
                    >
                      {errors.name}
                    </span>
                  }
                </div>


                <div className="form-group form-width">
                  <input
                    type="text"
                    id="recipeIngredients"
                    name="ingredients"
                    className="form-control"
                    placeholder="A comma seperated list of ingredients..."
                    value={newRecipe.ingredients}
                    onChange={this.onChange}
                  />
                  {
                    errors.ingredients &&
                    <span
                      className="help-block text-danger"
                    >
                      {errors.ingredients}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <textarea
                    type="text"
                    rows="5"
                    id="recipeDescription"
                    className="form-control"
                    placeholder="Instructions on how this dish is made..."
                    name="description"
                    value={newRecipe.description}
                    onChange={this.onChange}
                  />
                  {
                    errors.description &&
                    <span
                      className="help-block text-danger"
                    >
                      {errors.description}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <button
                    id="recipeButton"
                    className="btn btn-secondary"
                    disabled={this.state.submitting}
                  >
                    {this.state.submitting ? 'Submitting...' : 'Add Recipe'}
                  </button>
                  <Link
                    className="btn btn-secondary"
                    to="/users/recipes"
                  >
                    Cancel
                  </Link>
                </div>
              </form>
            </div>
          </div>
        </section>
      </div>

    );
  }
}

AddRecipe.propTypes = {
  addNewRecipe: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};


const mapStateToProps = state => ({
  addRecipe: state.addRecipe,
  authenticated: state.auth.isAuthenticated,
  errorMessage: state.errorMessage
});

const ConnectedComponent =
  connect(mapStateToProps, { addNewRecipe: addRecipe })(AddRecipe);
export { ConnectedComponent as ConnectedAddRecipe };

