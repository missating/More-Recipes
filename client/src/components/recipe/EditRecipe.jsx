import React from 'react';
import { connect } from 'react-redux';
import axios from 'axios';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';
import { Link, Redirect } from 'react-router-dom';

// validations
import recipeValidator from '../../validation/recipeValidator';

// action
import getSingleRecipe from '../../actions/getSingleRecipe';
import editRecipe from '../../actions/editRecipe';


/**
 * @description Creates EditRecipe component
 *
 * @class EditRecipe
 *
 * @extends {React.Component}
 */
export class EditRecipe extends React.Component {
  /**
   * @description Creates an instance of EditRecipe.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      name: '',
      ingredients: '',
      description: '',
      recipeImage: '',
      newImage: '',
      updating: false,
      errors: {},
    };
    this.onChange = this.onChange.bind(this);
    this.onEdit = this.onEdit.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
  }


  /**
   * @description Get's a recipe
   *
   * @method
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  componentDidMount() {
    const recipeId = this.props.match.params.id;
    this.props.getRecipeDetails(recipeId);
  }


  /**
   * @description Sets state
   *
   * @param {object} nextProps
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    const {
      id, name, description, ingredients, recipeImage
    } = nextProps.singleRecipe;
    this.setState({
      id, name, description, ingredients, recipeImage
    });
  }


  /**
   * @description Bind the value of the inputs to state
   *
   * @method onChange
   *
   * @param {object} event
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
  }


  /**
   * @description Update's the Recipe
   *
   * @method onEdit
   *
   * @param {object} event
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  onEdit(event) {
    event.preventDefault();
    this.setState({
      updating: true
    });
    const {
      id, name, description, ingredients, newImage, recipeImage
    } = this.state;
    const isValid = this.isValid();
    if (isValid) {
      if (newImage) {
        this.uploadToCloudinary().then((response) => {
          const secureUrl = response.data.secure_url;
          this.props.updateRecipe(
            {
              name, description, ingredients, recipeImage: secureUrl
            },
            id
          ).then(() => {
            this.props.history.push('/users/recipes');
          });
        });
      } else {
        this.props.updateRecipe(
          {
            name, description, ingredients, recipeImage
          },
          id
        ).then(() => {
          this.props.history.push('/users/recipes');
        });
      }
    }
  }

  /**
   * @description Handles image drop
   *
   * @method handleDrop
   *
   * @memberof EditRecipe
   *
   * @param {array} files
   *
   * @returns {void}
   */
  handleDrop(files) {
    this.setState({
      newImage: files[0],
      recipeImage: files[0].preview
    });
  }

  /**
   * @description Handle image upload
   *
   * @method uploadToCloudinary
   *
   * @memberof EditRecipe
   *
   * @returns {void}
   */
  uploadToCloudinary() {
    const formData = new FormData();
    formData.append('file', this.state.newImage);
    formData.append('upload_preset', 'moreRecipes');
    formData.append('api_key', '462882972372719');

    return axios.post(
      'https://api.cloudinary.com/v1_1/dxayftnxb/image/upload',
      formData, {
        headers: { 'X-Requested-With': 'XMLHttpRequest' },
      }
    );
  }

  /**
    * @description Validates user's data before making post request
    *
    * @method isValid
    *
    * @memberof EditRecipe
    *
    * @returns {boolean} true or false
    */
  isValid() {
    const { errors, isValid } = recipeValidator(this.state);
    if (!isValid) {
      this.setState({ errors });
    } else {
      this.setState({ errors: {} });
      return isValid;
    }
  }

  /**
    * @description Renders react component
    *
    * @method render
    *
    * @memberof EditRecipe
    *
    * @returns {void}
    *
    */
  render() {
    return (
      <div>
        {
          !this.props.authenticated &&
          <Redirect to="/" />
        }

        <section className="container" id="recipes" >

          <h3 className="text-center mb-4">Edit Recipe</h3>

          <div className="row">
            <div className="col-md-12 col-sm-12">
              <form>
                <Dropzone
                  onDrop={this.handleDrop}
                  accept="image/*"
                  multiple={false}
                  style={{
                    border: '3px dashed ',
                    width: '300px',
                    height: 'auto',
                    margin: '0 auto',
                    padding: '5px'
                  }}
                >
                  {
                    this.state.recipeImage &&
                    <div>
                      <img
                        src={this.state.recipeImage}
                        alt=""
                        className="img-fluid mx-auto d-block"
                      />
                      <h5 className="text-center">
                        Click here to upload
                      </h5>
                    </div>
                  }
                </Dropzone>

                <div className="form-group form-width">
                  <h5> Name </h5>
                  <input
                    type="text"
                    className="form-control"
                    name="name"
                    value={this.state.name}
                    onChange={this.onChange}
                  />
                  {
                    this.state.errors.name &&
                    <span
                      className="help-block text-danger"
                    >
                      {this.state.errors.name}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <h5> Ingredients </h5>
                  <input
                    type="text"
                    className="form-control"
                    name="ingredients"
                    value={this.state.ingredients}
                    onChange={this.onChange}
                  />
                  {
                    this.state.errors.ingredients &&
                    <span
                      className="help-block text-danger"
                    >
                      {this.state.errors.ingredients}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <h5> Description </h5>
                  <textarea
                    type="text"
                    rows="5"
                    className="form-control"
                    id="updateDescription"
                    name="description"
                    value={this.state.description}
                    onChange={this.onChange}
                  />
                  {
                    this.state.errors.description &&
                    <span
                      className="help-block text-danger"
                    >
                      {this.state.errors.description}
                    </span>
                  }
                </div>

                <div className="form-group form-width">
                  <button
                    id="editButton"
                    className="btn btn-secondary"
                    disabled={this.state.updating}
                    onClick={this.onEdit}
                  >
                    {this.state.updating ? 'Updating...' : 'Update'}
                  </button>

                  <Link
                    className="btn btn-secondary"
                    to="/users/recipes"
                  >
                    Cancel
                  </Link>
                </div>
              </form>
            </div>
          </div>
        </section >
      </div >
    );
  }
}


EditRecipe.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  singleRecipe: PropTypes.shape({}).isRequired,
  getRecipeDetails: PropTypes.func.isRequired,
  updateRecipe: PropTypes.func.isRequired,
  authenticated: PropTypes.bool.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};


const mapStateToProps = state => ({
  singleRecipe: state.recipes.singleRecipe,
  authenticated: state.auth.isAuthenticated
});

const mapDispatchToProps = dispatch => ({
  getRecipeDetails: id => dispatch(getSingleRecipe(id)),
  updateRecipe: (recipe, id) => dispatch(editRecipe(recipe, id))
});

const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(EditRecipe);
export { ConnectedComponent as ConnectedEditRecipe };

