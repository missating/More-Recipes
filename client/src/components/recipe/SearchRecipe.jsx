import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import searchRecipes from '../../actions/searchRecipe';


/**
 * @description Creates SearchRecipe component
 *
 * @class SearchRecipe
 *
 * @extends {React.Component}
 */
export class SearchRecipe extends React.Component {
  /**
  * @description Creates an instance of SearchRecipe.
  *
  * @constructor
  *
  * @param {object} props
  *
  * @memberof SearchRecipe
  *
  * @returns {void}
  */
  constructor(props) {
    super(props);
    this.state = {
      searchQuery: ''
    };
    this.onChange = this.onChange.bind(this);
  }

  /**
  * @description Bind the value of the inputs to state
  *
  * @method onChange
  *
  * @param {object} event
  *
  * @memberof SearchRecipe
  *
  * @returns {void}
  */
  onChange(event) {
    this.setState({ [event.target.name]: event.target.value });
    this.props.searchRecipes(
      event.target.value,
      this.props.pagination.currentPage
    );
  }

  /**
    * @description Renders react component
    *
    * @method render
    *
    * @memberof SearchRecipe
    *
    * @returns {void}
    *
    */
  render() {
    return (
      <div>
        <div className="col-lg-12">
          <form>
            <div className="input-group">
              <input
                id="search"
                type="text"
                name="searchQuery"
                className="form-control"
                value={this.state.searchQuery}
                onChange={this.onChange}
                placeholder="Search for..."
              />
            </div>
          </form>
        </div >
        <hr />
      </div >

    );
  }
}

SearchRecipe.propTypes = {
  searchRecipes: PropTypes.func.isRequired,
  pagination: PropTypes.shape({
    currentPage: PropTypes.number
  })
};

SearchRecipe.defaultProps = {
  pagination: {
    currentPage: 1
  }
};

const mapDispatchToProps = dispatch => ({
  searchRecipes: searchQuery => (dispatch(searchRecipes(searchQuery)))
});

const mapStateToProps = state => ({
  pagination: state.pagination,
});


const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(SearchRecipe);
export { ConnectedComponent as ConnectedSearchRecipe };
