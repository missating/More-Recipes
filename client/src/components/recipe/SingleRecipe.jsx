import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';


// actions
import getSingleRecipe from '../../actions/getSingleRecipe';
import addFavourite from '../../actions/addFavourite';
import voteRecipe from '../../actions/voteRecipe';

// components
import ActionButtons from '../common/ActionButtons';
import { ConnectedAddReview } from '../reviews/AddReview';
import ViewReviews from '../reviews/ViewReviews';

/**
 * @description Creates SingleRecipe component
 *
 * @class SingleRecipe
 *
 * @extends {React.Component}
 */
export class SingleRecipe extends React.Component {
  /**
   * @description Creates an instance of SingleRecipe.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof SingleRecipe
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
      singleRecipe: {},
      reviews: []
    };
    this.onFavourite = this.onFavourite.bind(this);
    this.onUpvote = this.onUpvote.bind(this);
    this.onDownVote = this.onDownVote.bind(this);
  }


  /**
   * @description Gets a single recipe
   *
   * @method
   *
   * @memberof SingleRecipe
   *
   * @returns {void}
   */
  componentDidMount() {
    const recipeId = this.props.match.params.id;
    this.props.recipe(recipeId);
  }

  /**
   * @description Set state
   *
   * @param {object} nextProps
   *
   * @memberof SingleRecipe
   *
   * @returns {void}
   */
  componentWillReceiveProps(nextProps) {
    const singleRecipe = (nextProps.singleRecipe) ? (
      nextProps.singleRecipe) : {};
    const reviews = (singleRecipe.Reviews) ? (singleRecipe.Reviews) : [];
    this.setState({ singleRecipe, reviews });
  }

  /**
    * @description Favourites a recipe with the it's Id
    *
    * @method onFavourite
    *
    * @memberof SingleRecipe
    *
    * @returns {void}
    */
  onFavourite() {
    const recipeId = this.props.singleRecipe.id;
    this.props.favourite(recipeId);
  }

  /**
    * @description Upvotes a recipe with the it's Id
    *
    * @method onUpvote
    *
    * @memberof SingleRecipe
    *
    * @returns {void}
    */
  onUpvote() {
    const recipeId = this.props.singleRecipe.id;
    this.props.upvote(recipeId, 'upvote');
  }

  /**
   * @description Downvotes a recipe with the it's Id
   *
   * @method onDownVote
   *
   * @memberof SingleRecipe
   *
   * @returns {void}
   */
  onDownVote() {
    const recipeId = this.props.singleRecipe.id;
    this.props.upvote(recipeId, 'downvote');
  }


  /**
     * @description Renders react component
     *
     * @method render
     *
     * @memberof SingleRecipe
     *
     * @returns {void}
     *
     */
  render() {
    const { singleRecipe } = this.props;

    if (singleRecipe.singleRecipe && !singleRecipe.singleRecipe.name) {
      return (
        <div className="loader-container">
          <div className="loader" />
        </div>
      );
    }

    const { reviews } = this.state;
    const allReviews = reviews.map((review, i) => (
      <div key={`review ${i + 1}`} className="container">
        <ViewReviews
          user={review.User.fullname}
          email={review.User.email}
          content={review.content}
          created={new Date(review.createdAt).toDateString()}
        />
      </div>
    ));

    return (
      <div>
        <section className="container" id="recipes">
          <div className="row">
            <div className="col-md-6 col-sm-12">
              <h2 className="ml-4">{this.props.singleRecipe.name}</h2>
              <img
                src={this.props.singleRecipe.recipeImage}
                alt={this.props.singleRecipe.name}
                className="img-thumbnail"
              />
              {
                !this.props.authenticated &&
                <ActionButtons singleRecipe={this.state.singleRecipe} />
              }
              {
                this.props.authenticated &&
                <div className="container">
                  <button
                    id="onDownvote"
                    className="btn active"
                    onClick={this.onDownVote}
                  >
                    <span>
                      <i className="far fa-thumbs-down" />
                    </span>
                    {this.props.singleRecipe.downvote}
                  </button>

                  <button
                    id="onUpvote"
                    className="btn active"
                    onClick={this.onUpvote}
                  >
                    <span>
                      <i className="far fa-thumbs-up" />
                    </span>
                    {this.props.singleRecipe.upvote}
                  </button>

                  <button
                    id="onFavourite"
                    className="btn active"
                    onClick={this.onFavourite}
                  >
                    <span>
                      <i className="far fa-heart" />
                    </span>
                    {this.props.singleRecipe.favourite}
                  </button>

                </div>
              }
            </div>
            <div className="col-md-6 col-sm-12 recipe-details">
              <hr />
              <h5>Ingredients</h5>
              <p>{this.props.singleRecipe.ingredients}</p>
              <hr />
              <h5> Method </h5>
              <p>{this.props.singleRecipe.description}</p>
            </div>
          </div>

          <div className="row mt-5 recipe-details">
            <div className="col-md-6 col-sm-12 reviews">
              <h4> Reviews </h4>
              {
                allReviews.length > 0 &&
                <div>
                  {allReviews}
                </div>
              }
              {
                !allReviews.length &&
                <div className="container">
                  <p>Currently no reviews</p>
                </div>
              }
            </div>
            <div className="col-md-6 col-sm-12">
              {this.props.authenticated &&
                <ConnectedAddReview recipeId={this.props.singleRecipe.id} />
              }
              {!this.props.authenticated &&
                <div>
                  <p className="ml-5"> Sign in to add a review </p>
                  <Link
                    className="btn btn-secondary ml-5"
                    to="/"
                  >
                    SIGN IN
                  </Link>
                </div>
              }
            </div>
          </div>
        </section>
      </div>
    );
  }
}

SingleRecipe.propTypes = {
  singleRecipe: PropTypes.shape({
    id: PropTypes.number.isRequired,
    recipeImage: PropTypes.string,
    name: PropTypes.string,
    ingredients: PropTypes.string,
    description: PropTypes.string,
    favourite: PropTypes.number,
    upvote: PropTypes.number,
    downvote: PropTypes.number
  }).isRequired,
  recipe: PropTypes.func.isRequired,
  upvote: PropTypes.func.isRequired,
  favourite: PropTypes.func.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      id: PropTypes.string.isRequired
    }).isRequired
  }).isRequired,
  authenticated: PropTypes.bool.isRequired,
};


const mapStateToProps = state => ({
  singleRecipe: state.recipes.singleRecipe,
  authenticated: state.auth.isAuthenticated,
  isFetching: state.isFetching
});

const mapDispatchToProps = dispatch => ({
  recipe: id => dispatch(getSingleRecipe(id)),
  favourite: recipeId => dispatch(addFavourite(recipeId)),
  upvote: (recipeId, queryType) => dispatch(voteRecipe(recipeId, queryType))
});

const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(SingleRecipe);
export { ConnectedComponent as ConnectedSingleRecipe };

