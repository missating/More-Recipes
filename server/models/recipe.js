export default (sequelize, DataTypes) => {
  const Recipe = sequelize.define('Recipe', {
    name: {
      type: DataTypes.STRING,
      allowNull: false
    },

    userId: {
      type: DataTypes.INTEGER,
      allowNull: false
    },

    ingredients: {
      type: DataTypes.STRING,
      allowNull: false
    },

    description: {
      type: DataTypes.TEXT,
      allowNull: false
    },

    recipeImage: {
      type: DataTypes.STRING,
      allowNull: true
    },


    favourite: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0
    },

  });

  Recipe.associate = (models) => {
    // associations defined here
    Recipe.belongsTo(
      models.User,
      { foreignKey: 'userId', onDelete: 'CASCADE', hooks: true }
    );
    Recipe.hasMany(
      models.Review,
      { foreignKey: 'recipeId', onDelete: 'CASCADE', hooks: true }
    );
    Recipe.hasMany(
      models.Vote,
      { foreignKey: 'recipeId', onDelete: 'CASCADE', hooks: true }
    );
    Recipe.hasMany(
      models.Favourite,
      { foreignKey: 'recipeId', onDelete: 'CASCADE', hooks: true }
    );
  };
  return Recipe;
};
