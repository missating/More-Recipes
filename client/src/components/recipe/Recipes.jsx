import React from 'react';
import { connect } from 'react-redux';
import ReactPaginate from 'react-paginate';
import PropTypes from 'prop-types';

// components
import RecipeCard from '../cards/RecipeCard';
import { ConnectedSearchRecipe } from '../recipe/SearchRecipe';

// actions
import getAllRecipes from '../../actions/getAllRecipes';

/**
 * @description Creates Recipes component
 *
 * @class Recipes
 *
 * @extends {React.Component}
 */
export class Recipes extends React.Component {
  /**
   * @description Creates an instance of Recipes.
   *
   * @constructor
   *
   * @param {object} props
   *
   * @memberof Recipes
   *
   * @returns {void}
   */
  constructor(props) {
    super(props);
    this.state = {
    };
    this.onPageChange = this.onPageChange.bind(this);
  }


  /**
   * @description Get recipes
   *
   * @memberof Recipes
   *
   * @returns {void}
   */
  componentWillMount() {
    this.props.recipes();
  }

  /**
   * @description pagination for recipes
   *
   * @param {number} current
   *
   * @memberof Recipes
   *
   * @returns {void}
   */
  onPageChange(current) {
    const selected = current.selected + 1;
    this.props.history.push(`/recipes?page=${selected}`);
    this.props.recipes(selected);
  }


  /**
   * @description Renders react component
   *
   * @method render
   *
   * @memberof Recipes
   *
   * @returns {void}
   *
   */
  render() {
    const { pages } = this.props.pagination;

    const recipes = this.props.allRecipes ? this.props.allRecipes : [];
    let recipeList;
    if (recipes.length === 0) {
      recipeList = (
        <div className="text-center ml-5" style={{ width: '100%' }}>
          <h4> No recipe(s) found </h4>
        </div>
      );
    } else {
      recipeList = recipes.map((recipe, i) => (
        <div className="col-md-6 col-lg-4 p-0" key={`recipe${i + 1}`}>
          <RecipeCard
            {...recipe}
          />
        </div>
      ));
    }


    return (
      <div>
        <section className="container" id="recipes">

          <ConnectedSearchRecipe />

          <div className="row mh-500">
            {recipeList}
          </div>


          <div className="container">
            <ReactPaginate
              pageCount={pages}
              pageRangeDisplayed={5}
              marginPagesDisplayed={3}
              previousLabel="Previous"
              nextLabel="Next"
              breakClassName="text-center"
              initialPage={0}
              containerClassName="container pagination justify-content-center"
              pageClassName="page-item"
              pageLinkClassName="page-link"
              activeClassName="page-item active"
              previousClassName="page-item"
              nextClassName="page-item"
              nextLinkClassName="page-link"
              previousLinkClassName="page-link"
              onPageChange={this.onPageChange}
            />
          </div>


        </section>
      </div>
    );
  }
}

Recipes.propTypes = {
  recipes: PropTypes.func.isRequired,
  allRecipes: PropTypes.arrayOf(PropTypes.object),
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  pagination: PropTypes.shape({
    pages: PropTypes.number
  })
};

Recipes.defaultProps = {
  pagination: {
    pages: 1
  },
  allRecipes: []
};

const mapStateToProps = state => ({
  allRecipes: state.recipes.allrecipes,
  pagination: state.pagination,
  isFetching: state.isFetching,
});

const mapDispatchToProps = dispatch => ({
  recipes: page => dispatch(getAllRecipes(page))
});

const ConnectedComponent =
  connect(mapStateToProps, mapDispatchToProps)(Recipes);
export { ConnectedComponent as ConnectedRecipes };

