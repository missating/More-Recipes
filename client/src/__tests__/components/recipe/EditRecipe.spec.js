import React from 'react';
import toJSON from 'enzyme-to-json';
import sinon from 'sinon';
import moxios from 'moxios';
import { EditRecipe } from '../../../components/recipe/EditRecipe';

describe('Edit Recipe component', () => {
  beforeEach(() => {
    moxios.install();
  });
  afterEach(() => {
    moxios.uninstall();
  });

  const props = {
    authenticated: true,
    match: {
      params: {
        id: '1'
      }
    },
    history: {
      push: jest.fn()
    },
    singleRecipe: {
      id: 1,
      name: 'test',
      ingredients: 'test, test',
      description: 'mix it',
      favourite: 0,
      upvote: 0,
      downvote: 0,
      Reviews: []
    },
    getRecipeDetails: jest.fn(() => Promise.resolve()),
    updateRecipe: jest.fn(() => Promise.resolve())
  };


  it('Should render', () => {
    const wrapper = shallow(<EditRecipe
      {...props}
    />);
    expect(wrapper.find('form').exists).toBeTruthy();
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  it(
    'calls onChange method when the value of the edit recipe input changes',
    () => {
      const spy = sinon.spy(EditRecipe.prototype, 'onChange');
      const wrapper = shallow(<EditRecipe
        {...props}
      />);
      wrapper.find('#updateDescription').simulate('change', {
        target: { name: 'description', value: 'recipe description' }
      });
      expect(spy.called).toBeTruthy();
    }
  );

  it('Should update recipe when submit button is clicked', () => {
    const spy = sinon.spy(EditRecipe.prototype, 'onEdit');
    const wrapper = shallow(<EditRecipe
      {...props}
    />);
    wrapper.find('#editButton').simulate(
      'click',
      { preventDefault: jest.fn() }
    );
    expect(spy.called).toBeTruthy();
  });


  it('Should add recipe when form is submitted with image', () => {
    moxios.stubRequest(
      'https://api.cloudinary.com/v1_1/dxayftnxb/image/upload',
      {
        status: 200,
        response: {
          imageUrl: 'image_url'
        }
      }
    );
    const wrapper = shallow(<EditRecipe
      {...props}
    />);
    wrapper.instance().setState({
      name: 'Test',
      ingredients: 'test, test',
      description: 'This is just a test recipe',
      recipeImage: {
        preview: 'preview_url'
      },
      newImage: {
        preview: 'preview_url'
      },
      submitting: false
    });
    wrapper.update();
    wrapper.find('#editButton').simulate(
      'click',
      { preventDefault: jest.fn() }
    );
    expect(wrapper.instance().state).toEqual({
      description: 'This is just a test recipe',
      errors: {},
      ingredients: 'test, test',
      name: 'Test',
      newImage: {
        preview: 'preview_url'
      },
      recipeImage: { preview: 'preview_url' },
      submitting: false,
      updating: true
    });
  });

  it('should set uploaded files to state', () => {
    const files = [
      {
        name: 'file 1',
        url: 'url'
      }
    ];
    const wrapper = shallow(<EditRecipe
      {...props}
    />);
    wrapper.instance().handleDrop(files);
    expect(wrapper.instance().state.newImage).toEqual({
      name: 'file 1', url: 'url'
    });
  });
});

