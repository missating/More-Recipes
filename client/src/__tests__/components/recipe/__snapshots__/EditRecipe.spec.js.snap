// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Recipe component Should render 1`] = `
<div>
  <section
    className="container"
    id="recipes"
  >
    <h3
      className="text-center mb-4"
    >
      Edit Recipe
    </h3>
    <div
      className="row"
    >
      <div
        className="col-md-12 col-sm-12"
      >
        <form>
          <t
            accept="image/*"
            disableClick={false}
            disablePreview={false}
            disabled={false}
            maxSize={Infinity}
            minSize={0}
            multiple={false}
            onDrop={[Function]}
            preventDropOnDocument={true}
            style={
              Object {
                "border": "3px dashed ",
                "height": "auto",
                "margin": "0 auto",
                "padding": "5px",
                "width": "300px",
              }
            }
          />
          <div
            className="form-group form-width"
          >
            <h5>
               Name 
            </h5>
            <input
              className="form-control"
              name="name"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="form-group form-width"
          >
            <h5>
               Ingredients 
            </h5>
            <input
              className="form-control"
              name="ingredients"
              onChange={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="form-group form-width"
          >
            <h5>
               Description 
            </h5>
            <textarea
              className="form-control"
              id="updateDescription"
              name="description"
              onChange={[Function]}
              rows="5"
              type="text"
              value=""
            />
          </div>
          <div
            className="form-group form-width"
          >
            <button
              className="btn btn-secondary"
              disabled={false}
              id="editButton"
              onClick={[Function]}
            >
              Update
            </button>
            <Link
              className="btn btn-secondary"
              replace={false}
              to="/users/recipes"
            >
              Cancel
            </Link>
          </div>
        </form>
      </div>
    </div>
  </section>
</div>
`;
